src/content/en/updates/posts/2016/03/web-enabled-internet-of-things.markdown

```diff
@@ -0,0 +1,523 @@
+<img alt="Error message" src="/web/updates/images/2016/03/web-enabled-internet-of-things/error.png" style="width: 50%; float: left"/>
+<br/>
+
+The right side shows an earlier version, show casing a simple error
+log that I added, easing the development.
+
+Web Bluetooth makes it dead easy to talk to a Bluetooth Low Energy
+device, so let's look at a simplified version of my connection code.
+If you don't know how promises work, please check out this resource
+before reading further:
+
+https://developers.google.com/web/fundamentals/primers/promises
+
+Connecting to a Bluetooth device involved a chain of promises.
+First we filter for the device (UUID: `FC00`, name: `Edison`), which
+will popup a dialog for the user to select the device given the
+filter. Then we will connect to the GATT service and get the primary
+service and then the characteristics, which we then act on (read
+values and set up notification callbacks).
+
+> NOTE: It is best practise to **not** fetch the characteristics
+> *in parallel* using something like `Promise.all([p1, p2])`, as that
+> will make sure the successive read/writes following in the promise
+> chain will happen properly.
+
+Below we show a simplified version of my code, which will only work
+with the latest Web Bluetooth API, thus requiring Chrome Dev on Android.
+
+{% highlight javascript %}
+navigator.bluetooth.requestDevice({
+  filters: [{ name: 'Edison' }],
+  optionalServices: [0xFC00]
+})
+
+.then((device) => {
+  return device.gatt.connect();
+})
+
+.then((server) => {
+  return server.getPrimaryService(0xFC00));
+});
+
+.then((service) => {
+  let p1 = () => service.getCharacteristic(0xFC0B)
+  .then((characteristic) => {
+    this.colorLedCharacteristic = characteristic;
+    this.readLedColor();
+   });
+
+  let p2 = () => service.getCharacteristic(0xFC0A)
+  .then((characteristic) => {
+    characteristic.addEventListener(
+      'Characteristicvaluechanged', this.onTemperatureChange);
+    return characteristic.startNotifications();
+  });
+
+  return p1().then(p2);
+})
+
+.catch((err) => {
+  // Catch any error.
+})
+          
+.then(() => {
+  // Connection fully established, unless there was an error above.
+});
+{% endhighlight %}
+
+Reading and writing a string from a `DataView`/`ArrayBuffer` (what
+the WebBluetooth API uses) is just as easy as using `Buffer` on the
+Node.JS side, all we need to use is `TextEncoder` and `TextDecoder`:
+
+{% highlight javascript %}
+readLedColor: function() {
+  return this.colorLedCharacteristic.readValue()
+  .then(function(data) {
+    // In Chrome 50+, a DataView is returned instead of an ArrayBuffer.
+    data = data.buffer ? data : new DataView(data);
+    let decoder = new TextDecoder("utf-8");
+    let decodedString = decoder.decode(data);
+    document.querySelector('#color').value = decodedString;
+  });
+},
+
+writeLedColor: function() {
```

Comment:
Not sure where you use this function but you should return the promise if you are planning on chaining other promises to it.